File: Python Version/order_manager.py
from order_book import OrderBook
from order_type import OrderType
from order_book_enums import OrderBookSide
from order import Order


class OrderManager:
    def __init__(self):
        self.order_books = {}

    def submit_order(
        self, symbol: str, order_type: OrderType, quantity: int, price: float
    ):
        order = Order(symbol, quantity, price, order_type)
        if symbol not in self.order_books:
            self.order_books[symbol] = OrderBook(symbol)
        self.order_books[symbol].add_order(order)
        print(f"Order submitted: {order}")

    def display_order_book(self, symbol: str):
        if symbol in self.order_books:
            order_book = self.order_books[symbol].get_order_book()
            print(f"Buy Orders for {symbol}:")
            for order in order_book[OrderBookSide.BID.value]:
                print(order)
            print(f"Sell Orders for {symbol}:")
            for order in order_book[OrderBookSide.ASK.value]:
                print(order)
        else:
            print(f"No orders found for {symbol}")

    def display_trades(self, symbol: str):
        if symbol in self.order_books:
            trades = self.order_books[symbol].get_trades()
            print(f"Trades for {symbol}:")
            for trade in trades:
                print(trade)
        else:
            print(f"No trades found for {symbol}")

File: C++ Version/src/OrderManager.cpp
#include "OrderManager.h"
#include <iostream>

OrderManager::OrderManager()
    : m_orderBooks{}
{}

void OrderManager::addOrder(
    std::string symbol, OrderType orderType, double price, int quantity
)
{
    std::string key{ symbol };
    if (m_orderBooks.find(key) == m_orderBooks.end())
    {
        m_orderBooks[key] = std::make_unique<OrderBook>(symbol);
    }

    Order newOrder{ symbol, price, quantity, orderType };
    m_orderBooks[key]->addOrder(newOrder);
    std::cout << "Order Added: " << newOrder.getRepresentation() << '\n';
}

void OrderManager::viewOrderBook(std::string_view symbol) const
{
    if (m_orderBooks.find(symbol.data()) == m_orderBooks.end())
    {
        std::cout << "Order book for " << symbol << " does not exist\n";
        return;
    }
    m_orderBooks.at(symbol.data())->displayOrderBook();
}

void OrderManager::viewTrades(std::string_view symbol) const
{
    if (m_orderBooks.find(symbol.data()) == m_orderBooks.end())
    {
        std::cout << "Order book for " << symbol << " does not exist\n";
        return;
    }
    m_orderBooks.at(symbol.data())->displayTrades();
}
